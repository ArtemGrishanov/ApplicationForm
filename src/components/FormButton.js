import React from 'react'

export default function FormButton({text = 'button', disabled = false, onClick}) {
    return <button className={"btn" + ((disabled) ? ' __disabled': '')} onClick={onClick}>{text}</button>
}
