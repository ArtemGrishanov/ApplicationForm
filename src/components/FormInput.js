import React from 'react'

export default function FormInput({disable = false, placeholder = '', value = '', onValueChange}) {
    return (
        <input className="formInput"
            placeholder={placeholder}
            defaultValue={value}
            maxLength="100"
            onChange={event => onValueChange(event.target.value)}>
        </input>
    );
}
