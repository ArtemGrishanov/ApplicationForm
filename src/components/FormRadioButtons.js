import React from 'react'

export default class RadioButtons extends React.Component {

    constructor(props) {
        super(props);
        this.onChange = this.onChange.bind(this);
    }

    onChange(e) {
        this.props.onValueChange(e.target.value);
    }

    render() {
        return (
            <form>
                <fieldset>
                    {this.props.options.map((item, index) =>
                        <div key={item.key}>
                            <input type="radio"
                                id={item.key}
                                name={this.props.fieldId}
                                value={item.key}
                                onChange={this.onChange}
                                defaultChecked={this.props.value === item.key}
                                />
                            <label htmlFor={item.key}>{item.text}</label>
                        </div>
                    )}
                </fieldset>
            </form>
        );
    }
}
