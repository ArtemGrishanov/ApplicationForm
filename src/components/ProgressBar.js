import React from 'react'

export default function ProgressBar({percentage = 50}) {
    return <div></div>;
}
