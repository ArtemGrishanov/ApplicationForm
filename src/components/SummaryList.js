import React from 'react'

export default function SummaryList({dataArray}) {
    return (
        <ul className="sum_ul">
            {dataArray.map((item, index) =>
                <li className="sum_i" key={item.title}>
                    <p className="sum_i_t">{item.title}</p>
                    <p className="sum_i_v">{item.value}</p>
                </li>
            )}
        </ul>
    )
}
