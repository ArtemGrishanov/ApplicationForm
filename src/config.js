import React from 'react'
import FormInput from './components/FormInput'
import FormRadioButtons from './components/FormRadioButtons'
import { validateEmail } from './helper'

/**
* Allows to set up a form structure for various cases or test
*/
export const TENANT_SCHEMA = [
    {
        title: 'First step title',
        data: [
            {
                fieldId: 'user_name',
                title: 'User Name',
                component: FormInput,
                placeholder: 'ex. Alex Brown'
            },
            {
                fieldId: 'user_email',
                title: 'User Email',
                component: FormInput,
                placeholder: 'myemail@gmail.com',
                validateFunction: validateEmail
            }
        ]
    },
    {
        title: 'Specify your salary',
        data: [
            {
                fieldId: 'user_salary',
                title: 'Salary',
                component: FormRadioButtons,
                options: [
                    {key: "0_1000", text: "0 - 1.000"},
                    {key: "1000_2000", text: "1.000 - 2.000"},
                    {key: "2000_3000", text: "2.000 - 3.000"},
                    {key: "3000_4000", text: "3.000 - 4.000"},
                    {key: "4000+", text: "Mehr als 4.000"}
                ]
            }
        ]
    }
];
