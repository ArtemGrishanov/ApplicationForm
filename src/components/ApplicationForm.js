import React from 'react'
import FormButton from './FormButton'
import FormInput from './FormInput'
import ProgressBar from './ProgressBar'

export default class ApplicationForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            maxStepsCount: props.schema.length,
            progress: 0,
            formData: new Map(), // stores inputted user data
            isCurrentStepCompleted: false // indicates, we can move next
        };
        this.onBackClick = this.onBackClick.bind(this);
        this.onNextClick = this.onNextClick.bind(this);
        this.onSubmitClick = this.onSubmitClick.bind(this);
    }

    onFormDataChange(fieldId, value) {
        this.state.formData.set(fieldId, value);
        this.setState({
            isCurrentStepCompleted: this.isStepCompleted(this.state.progress)
        });
    }

    /**
    * Checks current step is completed and user can move on
    */
    isStepCompleted(progress) {
        const isSummary = progress >= this.state.maxStepsCount;
        if (isSummary === false) {
            const schemaStep = this.props.schema[progress];
            for (let i = 0; i < schemaStep.data.length; i++) {
                const sdata = schemaStep.data[i];
                const fv = this.state.formData.get(sdata.fieldId);
                if (sdata.validateFunction && sdata.validateFunction(fv) === false) {
                    // can define a specific checking for some fields (ex. email)
                    return false;
                }
                else if (fv === null || fv === undefined || fv === '') {
                    // common field value check
                    return false;
                }
            }
            return true;
        }
        return true;
    }

    onSubmitClick() {
        //TODO submit action
        console.dir(this.state.formData);
        alert('Submitted!');
    }

    onBackClick() {
        const p = this.state.progress-1;
        this.setState({
            progress: p,
            isCurrentStepCompleted: this.isStepCompleted(p)
        });
    }

    onNextClick() {
        if (this.state.isCurrentStepCompleted === true) {
            const p = this.state.progress+1;
            this.setState({
                progress: p,
                isCurrentStepCompleted: this.isStepCompleted(p)
            });
        }
    }

    render() {
        const isSummary = this.state.progress >= this.state.maxStepsCount;
        const schemaStep = isSummary ? null: this.props.schema[this.state.progress];

        return (
            <div>
                <ProgressBar progress={this.state.progress / this.state.maxStepsCount}/>
                {isSummary === false &&
                    <div>
                        <p>{schemaStep.title}</p>
                        <div text={this.state.errorText}></div>
                        {schemaStep.data.map((item, index) => {
                            const TagName = item.component;
                            return (
                                <div key={item.title}>
                                    <p>{item.title}</p>
                                    <TagName {...item} value={this.state.formData.get(item.fieldId)} onValueChange={this.onFormDataChange.bind(this, item.fieldId)}>
                                    </TagName>
                                </div>
                            )
                        })}
                    </div>
                }
                {isSummary === true &&
                    <p>Summary</p>
                }
                {this.state.progress > 0 &&
                    <FormButton text="Back" onClick={this.onBackClick}/>
                }
                {isSummary === false &&
                    <FormButton disabled={!this.state.isCurrentStepCompleted} text="Next" onClick={this.onNextClick}/>
                }
                {isSummary === true &&
                    <FormButton text="Submit" onClick={this.onSubmitClick}/>
                }
            </div>
        );
    }
}
