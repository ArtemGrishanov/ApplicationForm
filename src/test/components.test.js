import React from 'react';
import FormButton from '../components/FormButton'
import FormInput from '../components/FormInput'
import ProgressBar from '../components/ProgressBar'
import SummaryList from '../components/SummaryList'
import FormRadioButtons from '../components/FormRadioButtons'
import ApplicationForm from '../components/ApplicationForm'
import { TENANT_SCHEMA } from '../config'

it('FormInput rendering...', () => {
    const copy = shallow(
        <FormInput disable={false} placeholder="plsh" value = "" onValueChange={() => {}}/>
    );
    expect(copy).toMatchSnapshot();
});

it('FormButton rendering...', () => {
    const copy = shallow(
        <FormButton text="Text" disabled={false} onClick={() => {}}/>
    );
    expect(copy).toMatchSnapshot();
});

it('ProgressBar rendering...', () => {
    const copy = shallow(
        <ProgressBar progress="66"/>
    );
    expect(copy).toMatchSnapshot();
});

it('FormRadioButtons rendering...', () => {
    const options = [
        {key: "k1", text: "v1"},
        {key: "k2", text: "v2"},
        {key: "k3", text: "v3"},
        {key: "k4", text: "v4"},
        {key: "k5", text: "v5"}
    ];
    const copy = shallow(
        <FormRadioButtons onValueChange={() => {}} fieldId="id" value="v1" options={options}/>
    );
    expect(copy).toMatchSnapshot();
});

it('SummaryList rendering...', () => {
    const dataArray = [
        {title: "k1", value: "v1"},
        {title: "k2", value: "v2"}
    ];
    const copy = shallow(
        <SummaryList dataArray={dataArray}/>
    );
    expect(copy).toMatchSnapshot();
});

it('ApplicationForm rendering', ()=> {
    const copy = shallow(
        <ApplicationForm schema={TENANT_SCHEMA} onSubmit={(data) => {
            console.log(data)
        }}/>
    );
    expect(copy).toMatchSnapshot();
});
