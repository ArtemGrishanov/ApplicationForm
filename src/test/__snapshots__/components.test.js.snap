// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationForm rendering 1`] = `
<div
  className="frm_wr"
>
  <div
    className="frm_cont"
  >
    <ProgressBar
      percentage={1}
    />
    <div
      className="frm_data mt20 __shown"
    >
      <p
        className="frm_ttl"
      >
        Personal Info
      </p>
      <div
        className="frm_data_i"
        key="Full Name"
      >
        <p
          className="frm_data_ttl"
        >
          Full Name
        </p>
        <FormInput
          component={[Function]}
          fieldId="user_name"
          onValueChange={[Function]}
          placeholder="Hans MÃ¼ller"
          title="Full Name"
        />
      </div>
      <div
        className="frm_data_i"
        key="Email"
      >
        <p
          className="frm_data_ttl"
        >
          Email
        </p>
        <FormInput
          component={[Function]}
          fieldId="user_email"
          onValueChange={[Function]}
          placeholder="me@example.com"
          title="Email"
          validateFunction={[Function]}
        />
      </div>
    </div>
    <div
      className="frm_btns"
    >
      <FormButton
        disabled={true}
        onClick={[Function]}
        text="Next"
      />
    </div>
  </div>
</div>
`;

exports[`FormButton rendering... 1`] = `
<button
  className="btn"
  onClick={[Function]}
>
  Text
</button>
`;

exports[`FormInput rendering... 1`] = `
<input
  className="formInput"
  defaultValue=""
  maxLength="100"
  onChange={[Function]}
  placeholder="plsh"
/>
`;

exports[`FormRadioButtons rendering... 1`] = `
<form>
  <div
    className="rb_i"
    key="k1"
  >
    <input
      className="form_radio"
      defaultChecked={true}
      id="k1"
      name="id"
      onChange={[Function]}
      type="radio"
      value="v1"
    />
    <label
      htmlFor="k1"
    >
      v1
    </label>
  </div>
  <div
    className="rb_i"
    key="k2"
  >
    <input
      className="form_radio"
      defaultChecked={false}
      id="k2"
      name="id"
      onChange={[Function]}
      type="radio"
      value="v2"
    />
    <label
      htmlFor="k2"
    >
      v2
    </label>
  </div>
  <div
    className="rb_i"
    key="k3"
  >
    <input
      className="form_radio"
      defaultChecked={false}
      id="k3"
      name="id"
      onChange={[Function]}
      type="radio"
      value="v3"
    />
    <label
      htmlFor="k3"
    >
      v3
    </label>
  </div>
  <div
    className="rb_i"
    key="k4"
  >
    <input
      className="form_radio"
      defaultChecked={false}
      id="k4"
      name="id"
      onChange={[Function]}
      type="radio"
      value="v4"
    />
    <label
      htmlFor="k4"
    >
      v4
    </label>
  </div>
  <div
    className="rb_i"
    key="k5"
  >
    <input
      className="form_radio"
      defaultChecked={false}
      id="k5"
      name="id"
      onChange={[Function]}
      type="radio"
      value="v5"
    />
    <label
      htmlFor="k5"
    >
      v5
    </label>
  </div>
</form>
`;

exports[`ProgressBar rendering... 1`] = `
<div
  className="prgs_cnt"
>
  <div
    className="prgs_fill"
    style={
      Object {
        "width": "50%",
      }
    }
  />
</div>
`;

exports[`SummaryList rendering... 1`] = `
<ul
  className="sum_ul"
>
  <li
    className="sum_i"
    key="k1"
  >
    <p
      className="sum_i_t"
    >
      k1
    </p>
    <p
      className="sum_i_v"
    >
      v1
    </p>
  </li>
  <li
    className="sum_i"
    key="k2"
  >
    <p
      className="sum_i_t"
    >
      k2
    </p>
    <p
      className="sum_i_v"
    >
      v2
    </p>
  </li>
</ul>
`;
